src/NodeServices/Interfaces.cs

```diff
@@ -188,4 +191,108 @@ internal interface IExtensionApplication
     }
+
+    //TODO should this live in DynamoServices?
+    //make these internal somehow?
+    public interface IExternalCodeCompletionProviderCore
+    {
+
+        /// <summary>
+        /// Already discovered variable types
+        /// </summary>
+        Dictionary<string, Type> VariableTypes { get; set; }
+
+        /// <summary>
+        /// Types that have already been imported into the scope
+        /// </summary>
+        Dictionary<string, Type> ImportedTypes { get; set; }
+
+        /// <summary>
+        /// Generate completion data for the specified text, while import the given types into the
+        /// scope and discovering variable assignments.
+        /// </summary>
+        /// <param name="code">The code to parse</param>
+        /// <param name="expand">Determines if the entire namespace should be used</param>
+        /// <returns>Return a list of IronPythonCompletionData </returns>
+        IExternalCodeCompletionData[] GetCompletionData(string code, bool expand = false);
+
+        /// <summary>
+        /// Try to generate a description from a typename
+        /// </summary>
+        /// <param name="stub">Everything before the last namespace or type name e.g. System.Collections in System.Collections.ArrayList</param>
+        /// <param name="item">Everything after the stub</param>
+        /// <param name="isInstance">Whether it's an instance or not</param>
+        string GetDescription(string stub, string item, bool isInstance);
+
+        bool MatchingEngine(string engineName);
+
+        //TODO think about this, more general?
+        void ImportStdLibrary(string dynamoCorePath);
+
+        void Log(string message);
+
+    }
+
+
+    /// <summary>
+    /// This interface is essentially the same as ICompletionData from AvalonEdit 
+    /// except it does not include any references to WPF.
+    /// </summary>
+    public interface IExternalCodeCompletionData
+    {
+        //TODO hmmmmmm what about image path or image instead?
+      
+        ExternalCodeCompletionType CompletionType  { get; }
+        //
+        // Summary:
+        //     Gets the text. This property is used to filter the list of visible elements.
+        string Text { get; }
+        //
+        // Summary:
+        //     The displayed content. This can be the same as 'Text', or a WPF UIElement if
+        //     you want to display rich content.
+        object Content { get; }
+        //
+        // Summary:
+        //     Gets the description.
+        object Description { get; }
+        //
+        // Summary:
+        //     Gets the priority. This property is used in the selection logic. You can use
+        //     it to prefer selecting those items which the user is accessing most frequently.
+        double Priority { get; }
+    }
+
+    public enum ExternalCodeCompletionType
+    {
+        NAMESPACE,
+        METHOD,
+        FIELD,
+        CLASS,
+        PROPERTY,
+        ENUM,
+        OTHER
+    };
```

Comment:
NIT: In .NET SNAKE_CASE is not the usual for naming enum values. Also, I don't think we are using that convention either. Please change to PascalCase.
Also, the usual convention is to have the "unknown" value (in this case `OTHER`) be 0, so it should go first in the list.

Reply:
The names are used to match the old interface... I could write a converter I guess with a dictionary map instead.